### PR TITLE
SimPO : Simple Preference Optimization with a Reference-Free Reward

### PR DESCRIPTION
**Pull request by nivibilla**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
#1037

#### Changelog
What are the changes made in this PR?
New training method, simPO : https://arxiv.org/html/2405.14734v1

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [ ] remove TODO
- [ ] possibly refactor to merge DPO and simPO
- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add unit tests for any new functionality
- [ ] update docstrings for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

**Comment by nivibilla** [comment]
Some open questions:

- The only real change is the loss function. And an omission of two lines in the trainer. So I feel like there is a lot of duplicated code, is it possible to combine dpo and simPO into one trainer and based on the config apply the loss?
- I wasn't sure but for simPO since there is no need for reference log probs this [chunk](https://github.com/pytorch/torchtune/blob/8c6e57746a87ab284162b37d8f5eed609058ca0a/recipes/lora_simPO_distributed.py#L631) can be omitted right?
- What is the best way to test correctness against the princeton nlp release version?

**Comment by nivibilla** [comment]
@kartikayk I will take that previous offer to review and help 😆

**Comment by pbontrager** [comment]
I wanted to followup here to help this from getting stuck.
- I believe we want to add this as its own loss and make this work with the DPO recipe. Maybe that could be given a more general name as you mentioned on the issue.
- For correctness, as a start, you should be able to basically test your unit test values on the simPO loss and the reference simPO loss and get the same outcomes. After you see that the individual components are the same, try and do a small training run on both code bases and you should see very similar loss curves. You can attach all of this verification to the PR description.

**Comment by ebsmothers** [comment]
Hi @nivibilla are you still working on adding this? Do @pbontrager's suggestions make sense? Happy to provide any additional suggestions as needed here

**Comment by nivibilla** [comment]
Hey @ebsmothers Im currently occupied with other stuff tbh. Not sure when I will get round to this. I can close this PR and let someone else do it if necessary
